File: Samples for Solid Edge/[C#]-Samples for Solid Edge/C#/SolidEdge.AddIn.EdgeBar/EdgeBarController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.ComTypes;
using System.Text;

namespace SolidEdge.AddIn.EdgeBar
{
    /// <summary>
    /// Controller class for ISolidEdgeBarEx.
    /// </summary>
    public class EdgeBarController : SolidEdgeFramework.ISEAddInEdgeBarEvents, IDisposable
    {
        private bool _disposed = false;
        private MyAddIn _myAddIn;
        private SolidEdgeFramework.ISolidEdgeBarEx _edgeBar;
        private Dictionary<IConnectionPoint, int> _connectionPointDictionary = new Dictionary<IConnectionPoint, int>();
        private Dictionary<IntPtr, EdgeBarPage> _edgeBarPageDictionary = new Dictionary<IntPtr, EdgeBarPage>();

        public EdgeBarController(MyAddIn myAddIn)
        {
            _myAddIn = myAddIn;
            _edgeBar = (SolidEdgeFramework.ISolidEdgeBarEx)_myAddIn.AddIn;

            HookEvents(_myAddIn.AddIn, typeof(SolidEdgeFramework.ISEAddInEdgeBarEvents).GUID);
        }

        ~EdgeBarController()
        {
            Dispose(false);
        }

        #region SolidEdgeFramework.ISEAddInEdgeBarEvents

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEAddInEdgeBarEvents.AddPage event.
        /// </summary>
        public void AddPage(object theDocument)
        {
            AddPage(theDocument, new SelectSetEdgeBarControl());
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEAddInEdgeBarEvents.IsPageDisplayable event.
        /// </summary>
        public void IsPageDisplayable(object theDocument, string EnvironmentCatID, out bool vbIsPageDisplayable)
        {
            vbIsPageDisplayable = true;
        }

        /// <summary>
        /// Implementation of SolidEdgeFramework.ISEAddInEdgeBarEvents.RemovePage event.
        /// </summary>
        public void RemovePage(object theDocument)
        {
            // We use the IUnknown pointer of the document as the dictionary key.
            IntPtr pUnk = Marshal.GetIUnknownForObject(theDocument);
            Marshal.Release(pUnk);

            // If we have an EdgeBarPage for the document, remove it.
            if (_edgeBarPageDictionary.ContainsKey(pUnk))
            {
                RemovePage(_edgeBarPageDictionary[pUnk]);
            }
        }

        #endregion

        #region EdgeBarController methods

        private EdgeBarPage AddPage(object theDocument, EdgeBarControl edgeBarControl)
        {
            EdgeBarPage edgeBarPage = null;
            IntPtr hWndPage = IntPtr.Zero;

            // We use the IUnknown pointer of the document as the dictionary key.
            IntPtr pUnk = Marshal.GetIUnknownForObject(theDocument);
            Marshal.Release(pUnk);

            // Only add a new EdgeBarPage if one hasn't already been added.
            if (!_edgeBarPageDictionary.ContainsKey(pUnk))
            {
                if (_myAddIn.ResourceAssembly != null)
                {
                    System.Reflection.Assembly resourceAssembly = _myAddIn.ResourceAssembly;

                    // If ResourceAssembly is null, default to the currently executing assembly.
                    if (resourceAssembly == null)
                    {
                        resourceAssembly = Assembly.GetExecutingAssembly();
                    }

                    hWndPage = new IntPtr(_edgeBar.AddPageEx(theDocument, resourceAssembly.Location, edgeBarControl.BitmapID, edgeBarControl.ToolTip, 2));
                }

                // ISolidEdgeBarEx.AddPage() may return null.
                if (!hWndPage.Equals(IntPtr.Zero))
                {
                    edgeBarPage = new EdgeBarPage(hWndPage, theDocument, edgeBarControl);
                }

                _edgeBarPageDictionary.Add(pUnk, edgeBarPage);
            }
            else
            {
                edgeBarPage = _edgeBarPageDictionary[pUnk];
            }

            return edgeBarPage;
        }

        private void RemovePage(EdgeBarPage edgeBarPage)
        {
            IntPtr hWndEdgeBarPage = edgeBarPage.Handle;

            // We use the IUnknown pointer of the document as the dictionary key.
            IntPtr pUnk = Marshal.GetIUnknownForObject(edgeBarPage.Document);
            Marshal.Release(pUnk);

            if (_edgeBarPageDictionary.ContainsKey(pUnk))
            {
                _edgeBarPageDictionary.Remove(pUnk);
                edgeBarPage.Dispose();

                _edgeBar.RemovePage(edgeBarPage.Document, hWndEdgeBarPage.ToInt32(), 0);
            }
        }

        #endregion

        #region IConnectionPoint helpers

        private void HookEvents(object eventSource, Guid eventGuid)
        {
            IConnectionPointContainer container = null;
            IConnectionPoint connectionPoint = null;
            int cookie = 0;

            container = (IConnectionPointContainer)eventSource;
            container.FindConnectionPoint(eventGuid, out connectionPoint);

            if (connectionPoint != null)
            {
                connectionPoint.Advise(this, out cookie);
                _connectionPointDictionary.Add(connectionPoint, cookie);
            }
        }

        private void UnhookAllEvents()
        {
            Dictionary<IConnectionPoint, int>.Enumerator enumerator = _connectionPointDictionary.GetEnumerator();
            while (enumerator.MoveNext())
            {
                enumerator.Current.Key.Unadvise(enumerator.Current.Value);
            }

            _connectionPointDictionary.Clear();
        }

        #endregion

        #region IDisposable

        public void Dispose()
        {
            Dispose(true);
        }

        public void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                if (disposing)
                {
                    Dictionary<IntPtr, EdgeBarPage>.Enumerator enumerator = _edgeBarPageDictionary.GetEnumerator();

                    while (enumerator.MoveNext())
                    {
                        RemovePage(enumerator.Current.Value);
                    }

                    _edgeBarPageDictionary.Clear();
                }

                _edgeBarPageDictionary = null;
                _edgeBar = null;
                _myAddIn = null;
                _disposed = true;
            }
        }

        #endregion
    }
}

File: Samples for Solid Edge/[VB]-Samples for Solid Edge/VB/SolidEdge.AddIn.EdgeBar/EdgeBarController.vb
Imports System.Reflection
Imports System.Runtime.InteropServices
Imports System.Runtime.InteropServices.ComTypes
Imports System.Text

Namespace SolidEdge.AddIn.EdgeBar
	''' <summary>
	''' Controller class for ISolidEdgeBarEx.
	''' </summary>
	Public Class EdgeBarController
		Implements SolidEdgeFramework.ISEAddInEdgeBarEvents, IDisposable

		Private _disposed As Boolean = False
		Private _myAddIn As MyAddIn
		Private _edgeBar As SolidEdgeFramework.ISolidEdgeBarEx
		Private _connectionPointDictionary As New Dictionary(Of IConnectionPoint, Integer)()
		Private _edgeBarPageDictionary As New Dictionary(Of IntPtr, EdgeBarPage)()

'INSTANT VB NOTE: The parameter myAddIn was renamed since it may cause conflicts with calls to static members of the user-defined type with this name:
		Public Sub New(ByVal myAddIn_Renamed As MyAddIn)
			_myAddIn = myAddIn_Renamed
			_edgeBar = CType(_myAddIn.AddIn, SolidEdgeFramework.ISolidEdgeBarEx)

			HookEvents(_myAddIn.AddIn, GetType(SolidEdgeFramework.ISEAddInEdgeBarEvents).GUID)
		End Sub

		Protected Overrides Sub Finalize()
			Dispose(False)
		End Sub

		#Region "SolidEdgeFramework.ISEAddInEdgeBarEvents"

		''' <summary>
		''' Implementation of SolidEdgeFramework.ISEAddInEdgeBarEvents.AddPage event.
		''' </summary>
		Public Sub AddPage(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEAddInEdgeBarEvents.AddPage
			AddPage(theDocument, New SelectSetEdgeBarControl())
		End Sub

		''' <summary>
		''' Implementation of SolidEdgeFramework.ISEAddInEdgeBarEvents.IsPageDisplayable event.
		''' </summary>
		Public Sub IsPageDisplayable(ByVal theDocument As Object, ByVal EnvironmentCatID As String, <System.Runtime.InteropServices.Out()> ByRef vbIsPageDisplayable As Boolean) Implements SolidEdgeFramework.ISEAddInEdgeBarEvents.IsPageDisplayable
			vbIsPageDisplayable = True
		End Sub

		''' <summary>
		''' Implementation of SolidEdgeFramework.ISEAddInEdgeBarEvents.RemovePage event.
		''' </summary>
		Public Sub RemovePage(ByVal theDocument As Object) Implements SolidEdgeFramework.ISEAddInEdgeBarEvents.RemovePage
			' We use the IUnknown pointer of the document as the dictionary key.
			Dim pUnk As IntPtr = Marshal.GetIUnknownForObject(theDocument)
			Marshal.Release(pUnk)

			' If we have an EdgeBarPage for the document, remove it.
			If _edgeBarPageDictionary.ContainsKey(pUnk) Then
				RemovePage(_edgeBarPageDictionary(pUnk))
			End If
		End Sub

		#End Region

		#Region "EdgeBarController methods"

		Private Function AddPage(ByVal theDocument As Object, ByVal edgeBarControl As EdgeBarControl) As EdgeBarPage
			Dim edgeBarPage As EdgeBarPage = Nothing
			Dim hWndPage As IntPtr = IntPtr.Zero

			' We use the IUnknown pointer of the document as the dictionary key.
			Dim pUnk As IntPtr = Marshal.GetIUnknownForObject(theDocument)
			Marshal.Release(pUnk)

			' Only add a new EdgeBarPage if one hasn't already been added.
			If Not _edgeBarPageDictionary.ContainsKey(pUnk) Then
				If _myAddIn.ResourceAssembly IsNot Nothing Then
					Dim resourceAssembly As System.Reflection.Assembly = _myAddIn.ResourceAssembly

					' If ResourceAssembly is null, default to the currently executing assembly.
					If resourceAssembly Is Nothing Then
						resourceAssembly = System.Reflection.Assembly.GetExecutingAssembly()
					End If

					hWndPage = New IntPtr(_edgeBar.AddPageEx(theDocument, resourceAssembly.Location, edgeBarControl.BitmapID, edgeBarControl.ToolTip, 2))
				End If

				' ISolidEdgeBarEx.AddPage() may return null.
				If Not hWndPage.Equals(IntPtr.Zero) Then
					edgeBarPage = New EdgeBarPage(hWndPage, theDocument, edgeBarControl)
				End If

				_edgeBarPageDictionary.Add(pUnk, edgeBarPage)
			Else
				edgeBarPage = _edgeBarPageDictionary(pUnk)
			End If

			Return edgeBarPage
		End Function

		Private Sub RemovePage(ByVal edgeBarPage As EdgeBarPage)
			Dim hWndEdgeBarPage As IntPtr = edgeBarPage.Handle

			' We use the IUnknown pointer of the document as the dictionary key.
			Dim pUnk As IntPtr = Marshal.GetIUnknownForObject(edgeBarPage.Document)
			Marshal.Release(pUnk)

			If _edgeBarPageDictionary.ContainsKey(pUnk) Then
				_edgeBarPageDictionary.Remove(pUnk)
				edgeBarPage.Dispose()

				_edgeBar.RemovePage(edgeBarPage.Document, hWndEdgeBarPage.ToInt32(), 0)
			End If
		End Sub

		#End Region

		#Region "IConnectionPoint helpers"

		Private Sub HookEvents(ByVal eventSource As Object, ByVal eventGuid As Guid)
			Dim container As IConnectionPointContainer = Nothing
			Dim connectionPoint As IConnectionPoint = Nothing
			Dim cookie As Integer = 0

			container = CType(eventSource, IConnectionPointContainer)
			container.FindConnectionPoint(eventGuid, connectionPoint)

			If connectionPoint IsNot Nothing Then
				connectionPoint.Advise(Me, cookie)
				_connectionPointDictionary.Add(connectionPoint, cookie)
			End If
		End Sub

		Private Sub UnhookAllEvents()
			Dim enumerator As Dictionary(Of IConnectionPoint, Integer).Enumerator = _connectionPointDictionary.GetEnumerator()
			Do While enumerator.MoveNext()
				enumerator.Current.Key.Unadvise(enumerator.Current.Value)
			Loop

			_connectionPointDictionary.Clear()
		End Sub

		#End Region

		#Region "IDisposable"

		Public Sub Dispose() Implements IDisposable.Dispose
			Dispose(True)
		End Sub

		Public Sub Dispose(ByVal disposing As Boolean)
			If Not _disposed Then
				If disposing Then
					Dim enumerator As Dictionary(Of IntPtr, EdgeBarPage).Enumerator = _edgeBarPageDictionary.GetEnumerator()

					Do While enumerator.MoveNext()
						RemovePage(enumerator.Current.Value)
					Loop

					_edgeBarPageDictionary.Clear()
				End If

				_edgeBarPageDictionary = Nothing
				_edgeBar = Nothing
				_myAddIn = Nothing
				_disposed = True
			End If
		End Sub

		#End Region
	End Class
End Namespace
